Implement suggested follow-up questions

cassidyxu opened the pull request:
## Overview

Provide suggested follow-up questions after getting a response from the chatbot.

## Changes Made

### Chatbot

- Added suggested follow-up question buttons after every response from the chatbot
    - This takes the last prompt and response as input and uses OpenAI API’s text generation to suggest follow-ups based on the context
- Buttons for sample questions and follow-ups
    - left-aligned text and removed green circle icons
- Changed model from gpt-4-turbo-preview to gpt-4o-2024-05-13

## Test Coverage

- Tested to make sure the follow-up question buttons disappear after receiving a response from either:
    - clicking one of the follow-up question buttons
    - asking a new question through text input box
    - clicking one of the sample questions at the top
- Tested to make sure the follow-up question buttons remain even after user submits feedback on the response

## Screenshots
<img width="700" alt="Screenshot 2024-05-25 at 4 39 47 PM" src="https://github.com/Brainana/LexBudget/assets/68167806/ec827af6-fc2d-4ea2-9f8b-e681539c0dbd">

cassidyxu committed:
Implement suggested follow-up questions after receiving a response from the chatbot

ccurme commented on Chatbot.py:
```diff
@@ -428,6 +428,40 @@ async def runAgent(userQuery, chat_history):
 
+if 'clicked' not in st.session_state:
+    st.session_state.clicked = False
+clickedQuestion = None
```
is this variable used?

ccurme commented on Chatbot.py:
```diff
@@ -428,6 +428,40 @@ async def runAgent(userQuery, chat_history):
+    st.session_state.clicked_question = None
+
+def click_button(question):
+    print("button clicked!" + question)
```
might want whitespace separating the question, e.g.,
```python
print(f"button clicked! {question}")
```

ccurme commented on Chatbot.py:
```diff
@@ -428,6 +428,40 @@ async def runAgent(userQuery, chat_history):
+
+    follow_up_suggestions = client.chat.completions.create(
+        messages=[{"role": "user", "content": follow_up_query}],
+        model="gpt-4-turbo-preview",
```
Consider pinning to a specific model, e.g., `"gpt-4-turbo-2024-04-09"`.

You might try `"gpt-4o-2024-05-13"`, which is faster and cheaper (but maintains good performance).

ccurme commented on Chatbot.py:
```diff
@@ -428,6 +428,40 @@ async def runAgent(userQuery, chat_history):
+def suggestFollowUps():
+    follow_up_query = f"""
+    Given this chat history {chat_history}, Suggest 2 follow-up questions the user 
+    might ask next. Give your answer in the format where each line has a separate 
+    question. Do not number your responses."""
```
I left this comment earlier but github may have dropped it: consider using OpenAI's tool calling features for something like this, which lets you coerce the response to a desired schema and avoid prompting this way / parsing out the list later.

Example (using LangChain):
```python
from typing import List

from langchain_core.pydantic_v1 import BaseModel
from langchain_openai import ChatOpenAI


class FollowUpQuestions(BaseModel):
    """Follow up questions."""
    questions: List[str]

llm = ChatOpenAI(model="gpt-4o-2024-05-13", temperature=0)
structured_llm = llm.with_structured_output(FollowUpQuestions)
response = structured_llm.invoke(follow_up_query)
response.questions  # <-- will be a list of strings
```
Under the hood this just wraps OpenAI's tool calling features, which you can also access via their sdk: https://platform.openai.com/docs/guides/function-calling

cassidyxu committed:
Ensure follow-up buttons remain visible after providing feedback and modify follow-ups to use OpenAI's tool calling features
- remove print statements
- remove unused variables
- remove green circle icons in buttons
- change button text to be left-aligned
- keep follow-up buttons on page after submitting feedback
- change model from gpt-4-turbo-preview to gpt-4o-2024-05-13
- use tool calling to coerce follow-up response to a desired schema
- rename variables for clarity
- modify follow ups to only consider last asked question instead of entire chat history

cassidyxu committed:
Fix bin files

ccurme commented:
Looks good!
